### PR TITLE
Serve multistream graph captures from correct pool

### LINKED ISSUE #114320: When cuda graph captures multi-stream function, allocations are not serviced from the correct pool
Repro:
```
import torch
print(torch.__version__)

torch.cuda.empty_cache()
x=torch.randn(10240000, device="cuda")
y=torch.rand_like(x)
g=torch.cuda.CUDAGraph()
s0 = torch.cuda.Stream()
s1 = torch.cuda.Stream()
s0.wait_stream(torch.cuda.current_stream())
with torch.cuda.stream(s0):
  g.capture_begin()
  z = x+y
  with torch.cuda.stream(s1):
    s1.wait_stream(s0)
    w = z+y
  s0.wait_stream(s1)
  g.capture_end()
segments = torch.cuda.memory_snapshot()
for s in segments:
  print("POOL ID", s["segment_pool_id"], s["active_size"], s["total_size"], s["stream"])
```
produces
```
2.1.0+cu118
POOL ID (0, 0) 41943040 41943040 94099410458288 # w, not from pool, bug!
POOL ID (1, 0) 41943040 41943040 94099410421936 # z, from pool, ok
POOL ID (0, 0) 41943040 41943040 0 # x or y
POOL ID (0, 0) 1024 2097152 94099410421936 # something or other rng state related for graphs
POOL ID (0, 0) 41943040 41943040 0 # x or y
```
This is due to #96542 which changed the logic to determine if current allocation should be served from the pool. After #96542 only allocations happening on the original stream of the pool capture will be served from the pool. 
cc @mcarilli @ezyang @zdevito

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114647

This fixes #114320 by placing the logic for determining whether to allocate
to a pool inside a callback that is controlled by CUDAGraph.cpp or by the
python bound api to allocate a stream directly to a pool.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

### CODE REVIEW

**Review by ngimel — APPROVED** [review]
Thanks!

**Comment by ngimel** [comment]
would it make sense to add a test?

**Comment by eellison** [comment]
doing review this morning sorry for delay

File: aten/src/ATen/cuda/CUDAGraph.cpp [hunk 1]
```diff
@@ -136,7 +136,12 @@ void CUDAGraph::capture_begin(MempoolId_t pool/*=0*/, cudaStreamCaptureMode capt
   // Addendum: beginAllocateStreamToPool is now called before cudaStreamBeginCapture to prevent an
   // autograd thread's free() call triggering an invalid cudaEventRecord in the caching allocator
   // due to the capture status being updated _after_ a capture had already started.
-  c10::cuda::CUDACachingAllocator::beginAllocateStreamToPool(capture_dev_, capture_stream_, mempool_id_);
+  c10::cuda::CUDACachingAllocator::beginAllocateToPool(capture_dev_, mempool_id_, [this](cudaStream_t stream) {
+      cudaStreamCaptureStatus status;
+      CaptureId_t stream_capture_id;
+      AT_CUDA_CHECK(cudaStreamGetCaptureInfo(stream, &status, &stream_capture_id));
+      return status == cudaStreamCaptureStatus::cudaStreamCaptureStatusActive && stream_capture_id == capture_id_;
```
Comment (eellison) [inline aten/src/ATen/cuda/CUDAGraph.cpp:143]: So just to confirm, 

```
with torch.cuda.stream(s0):
  g.capture_begin()
  z = x+y
  with torch.cuda.stream(s1):
    s1.wait_stream(s0)
    w = z+y
  s0.wait_stream(s1)
```

For both the `z` and `w` allocation we will get the same `capture_id_` ?

File: aten/src/ATen/cuda/CUDAGraph.cpp [hunk 2]
```diff
@@ -136,7 +136,12 @@ void CUDAGraph::capture_begin(MempoolId_t pool/*=0*/, cudaStreamCaptureMode capt
   // Addendum: beginAllocateStreamToPool is now called before cudaStreamBeginCapture to prevent an
   // autograd thread's free() call triggering an invalid cudaEventRecord in the caching allocator
   // due to the capture status being updated _after_ a capture had already started.
-  c10::cuda::CUDACachingAllocator::beginAllocateStreamToPool(capture_dev_, capture_stream_, mempool_id_);
+  c10::cuda::CUDACachingAllocator::beginAllocateToPool(capture_dev_, mempool_id_, [this](cudaStream_t stream) {
+      cudaStreamCaptureStatus status;
+      CaptureId_t stream_capture_id;
+      AT_CUDA_CHECK(cudaStreamGetCaptureInfo(stream, &status, &stream_capture_id));
```
Comment (eellison) [inline aten/src/ATen/cuda/CUDAGraph.cpp:142]: We're calling `cudaStreamGetCaptureInfo` a lot more now, but that probably doesn't really matter, and this is off the hot path anyway

**Review by eellison — APPROVED** [review]
looks good - won't necessarily cover the non capturing case for when we are doing warmup but nonetheless we should land

File: aten/src/ATen/cuda/CUDAGraph.cpp (hunk 2 above)
Comment (ngimel) [reply aten/src/ATen/cuda/CUDAGraph.cpp:142]: We call it only when there are captures underway, the way we used to

File: aten/src/ATen/cuda/CUDAGraph.cpp (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/cuda/CUDAGraph.cpp:143]: Yes

**Comment by ngimel** [comment]
Should this be landed?

**Comment by zdevito** [comment]
I wanted to add a test that checks this behavior keeps working, but I haven't had time to write it yet.

**Comment by eellison** [comment]
land ?
